Practicum/Pract. 06/Практикум-6 - задача 2.cpp

```
@@ -0,0 +1,213 @@
+﻿// Практикум-6.cpp : This file contains the 'main' function. Program execution begins and ends there.
+//
+#include <iostream>
+using namespace std;
+
+class NumberSeries
+{
+private:
+    int* series = nullptr;
+    int size;
+    int capacity;
+    int initial;
+    int (*func)(int);
+
+    void free()
+    {
+        delete[] series;
+        series = nullptr;
+    }
+
+    void CopyFrom(const NumberSeries& other)
+    {
+        series = new int[capacity + 1];
+        for (int i = 0; i < size; i++) 
+        {
+            series[i] = other.series[i];
+        }
+        size = other.size;
+        capacity = other.capacity;
+        initial = other.initial;
+        func = other.func;
+    }
+
+    void resizeArray() 
+    {
+        int newCapacity = capacity * 2;
+        int* newSeries = new int[newCapacity];
+        for (int i = 0; i < size; i++) 
+        {
+            newSeries[i] = series[i];
+        }
+        free();
+        capacity = newCapacity;
+    }
+
+    void reset()
+    {
+        free();
+        capacity = 10;
+        size = 0;
+        series = new int[capacity];
+
+        if (func != nullptr) 
+        {
+            series[0] = initial;
+            size = 1;
+        }
+    }
+
+public:
+
+
+    NumberSeries(int a0 = 0, int capacity = 10, int (*func)(int) = nullptr) : initial(a0), capacity(capacity), func(func), size((func != nullptr) ? 1 : 0)
+    {
+        series = new int[capacity];
+        if (func != nullptr) 
+        {
+            series[0] = initial;
+        }

```

Why do you allow there to be series with invalid function? It's better to throw an exception when the function pointer is nullptr